### PR TITLE
[datalink] updated bluegiga code - Implemented check if full message …

### PR DESCRIPTION
**Pull request by kirkscheper**
…fits in buffer, tried to use funciton inputs instead of global in case the plan is to define the periph somewhere else in the future, removed some funcitons that were not being used, moved some funciton definitions to .c. Tested with LadyLisaBartBlueGiga

### CODE REVIEW

**Comment by kirkscheper** [comment]
This addresses issue #1338

File: sw/airborne/subsystems/datalink/bluegiga.c [hunk 1]
```diff
@@ -49,34 +49,71 @@ struct spi_transaction bluegiga_spi;
...
-static int dev_check_free_space(struct bluegiga_periph *p __attribute__((unused)), uint8_t len __attribute__((unused)))
+void bluegiga_transmit(struct bluegiga_periph *p, uint8_t data);
+void bluegiga_receive(struct spi_transaction *trans);
+
+// Functions for the generic link device device API
+// TODO: I tried to remove the hard coded bluegiga_p here but I clearly don't understand how the link_device calls are made
+static int dev_check_free_space(struct link_device *dev __attribute__((unused)), uint8_t len)
 {
-  return TRUE;
+  struct bluegiga_periph *p = &bluegiga_p; //(struct bluegiga_periph*)(dev->periph);
```
Comment (gautierhattenberger) [inline sw/airborne/subsystems/datalink/bluegiga.c]: why the cast is commented ? it's not working ?

**Comment by kirkscheper** [comment]
Yea, in short the cast didn't work. 

It is a bit confusing where the DefaultDevice or DOWNLINK_DEVICE (from downlink.h) is set but I guess it is not set correctly. Do you have any idea where the DOWNLINK_DEVICE should be defined. I can of course define it in the bluegiga.h as the link_device from the local variable 'bluegiga_p.device' but not sure if that is the idea...

**Comment by gautierhattenberger** [comment]
can you my fix: https://github.com/gautierhattenberger/paparazzi/commit/136969563a1422e426ce319526a29fc21dcd9e47 ?

**Comment by kirkscheper** [comment]
I tried this as well, got a seg fault. I will ask @dewagter about this.

**Comment by gautierhattenberger** [comment]
If you run your code with GDB, check that p->device.periph really points to the bluegiga_p structure, because it is this pointer that is passed to the generic link_device functions (see pprz_transport.c file for instance). That's why you should directly have your bluegiga struct pointer as first argument of the functions like dev_check_free_space (hence no cast needed, actually it was really wrong to do that here).

**Comment by kirkscheper** [comment]
I also don't understand that. For example, autopilot.c:205 static void send_fp(struct transport_tx *trans, struct link_device *dev) calls pprz_msg_send_ROTORCRAFT_FP (in messages.h: 2738) which in term send the link_device not the periph to the functions like put_bytes.

I had already briefly checked with GDB, the pointer sent to the put_byte function was not the dev.

**Comment by kirkscheper** [comment]
ah, I see in pprz_transport.c where it is dereferenced. but I still get seg faults. I will look into this tomorrow and post in more detail what is going on.

**Comment by gautierhattenberger** [comment]
in messages.h, only functions from the generic transport interface (see transport.h) are used, and they require a `link_device` struct as second argument (in general). That's why you only have `dev` here.
But in the implementation of the transports (pprz_transport, xbee, ...) where the generic `link_device` functions are actually called, it is `dev->periph` that should be passed.
So dev->periph` should point to your actual device structure (`bluegiga_p` in your case). But in `put_bytes` it should be `bluegiga_p->device`, the generic interface.
I agree that all this stuff is a bit painful to follow...

**Comment by kirkscheper** [comment]
I guess I had a typo somewhere before cause it works now... Better this way then.

**Pull request merged**